sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -2038,6 +2038,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val REPARTITION_BEFORE_INSERT =
+    buildConf("spark.sql.execution.repartitionBeforeInsert")
```

Comment:
I am not sure. We can already do this via specifying the hints, and the performance improvement doesn't look common enough to me.

Reply:
Because we cannot prevent users from using inappropriate columns for partition columns, such as primary keys.
This will generate too many files and affect the stability of NameNode and the entire cluster.

![image](https://user-images.githubusercontent.com/5399861/77612239-9bd30f00-6f62-11ea-9178-3bcd65aa4034.png)

Reply:
> I am not sure. We can already do this via specifying the hints, and the performance improvement doesn't look common enough to me.

As the platform becomes better and better, users' understanding of the underlying service becomes less and less,  it is difficult to control users' behavior and we'd better to make the engine smarter and more applicable.

Reply:
If this is better in 90% case, I am good to go. If it's better in 50% cases, is it worthwhile?

Reply:
> If this is better in 90% and worse in 10% cases , I might be okay. If it's better in 50% and worse in 50% cases, is it worthwhile?

For dynamic partition write, file size is `(shuffle partition size) * (table partition size)`.  after repartition file size is ` (table partition size)`, add `RepartitionByExpression` shuffle data is quick since without other computation.  we should concern is data skew, with AQE we can control each partition's size to match expected file size. 
IMO, if this pr can give a test case of data skew and how it behavior with AQE is better.